File: examples/object-oriented-programming/python/oop.py
class Box:

    def __init__(self, width, height):

        self.width = width
        self.height = height
        self.color = "Red"
        self.__private_variable = "PRIVATE TEXT"

    # Replace .description() with __str__()
    def __str__(self):

        return "(w:" + str(self.width) + ",h:" + str(self.height) + ":" + self.__private_variable + ")"

    # instance methods

    def get_area(self):

        return self.__calculate_area()

    def get_perimeter(self):

        perimeter = (self.width + self.height) * 2

        return perimeter

    def __calculate_area(self):

        area = self.width * self.height

        return area


box = Box(10, 20)

print("New box: " + str(box))

box.version = 1

width = box.width
height = box.height
color = box.color
version = box.version
area = box.get_area()
perimeter = box.get_perimeter()

print("Width of the box is : " + str(width))
print("Height of the box is : " + str(height))
print("Color of the box is : " + str(color))
print("Area of the box is : " + str(area))
print("Perimeter of the box is : " + str(perimeter))
print("Version of the box is : " + str(box.version))

class Cube(Box):

    def __init__(self, height, width, depth):

        self.width = width
        self.height = height
        self.depth = depth
        self.color = "Blue"

    def __str__(self):

        return "(w:" + str(self.width) + ",h:" + str(self.height) + ",d:" + str(self.depth) + ")"


    def get_volume(self):

        volume = self.width * self.height * self.depth

        return volume

    # overwrite method of parent class
    def get_area(self):

        return 9000

cube = Cube(10, 20, 30)

print("\nNew cube: " + str(cube))

width = cube.width
height = cube.height
depth = cube.depth
color = cube.color
area = cube.get_area()
perimeter = cube.get_perimeter()
volume = cube.get_volume()

print("Width of the cube is : " + str(width))
print("Height of the cube is : " + str(height))
print("Depth of the cube is : " + str(depth))
print("Color of the cube is : " + str(color))
print("Area of the cube is : " + str(area))
print("Perimeter of the cube is : " + str(perimeter))
print("Volume of the cube is : " + str(volume))

File: examples/object-oriented-programming/java/Oop.java
class Box {

  // class variables

  int width;
  int height;
  String color;
  static int version = 0;

  // class methods

   public int get_area() {

     return calculate_area();
   }

   int get_perimeter() {

     return calculate_perimeter();
   }

   int get_version() {

     return this.version;
   }

   String print_box() {

     return "(w:" + this.width + ",h:" + this.height + ")";
   }

   static String print_static_text() {

     return "print static text";
   }

   protected int calculate_area() {

      return (this.width * this.height);
   }

   private int calculate_perimeter() {

      return (this.width + this.height) * 2;
   }

   // constructor

   public Box(int width, int height) {

     this.width = width;

     this.height = height;

     this.color = "Red";
   };
}

// inheritance

class Cube extends Box {

  int depth;

  public Cube(int width, int height, int depth) {

    super(width, height);

    this.depth = depth;

    this.color = "Blue";
  }

  int get_volume() {

    int area = super.get_area();

    return (area * this.depth);
  }

  String print_cube() {

    return "(w:" + this.width + ",h:" + this.height + ",d:" + this.depth + ")";
  }
}

public class Oop {

  public static void main(String[] args) {

    Box box = new Box(10, 20);

    System.out.printf("New box: %s%n", box.print_box());

    box.version = 1;

    System.out.printf("Width of the box is : %d%n", box.width);
    System.out.printf("Height of the box is : %d%n", box.height);
    System.out.printf("Color of the box is : %s%n", box.color);
    System.out.printf("Area of the box is : %d%n", box.get_area());
    System.out.printf("Perimeter of the box is : %d%n", box.get_perimeter());
    System.out.printf("Version of the box is : %d%n", box.get_version());
    System.out.printf("Static text of the box is : %s%n", box.print_static_text());

    Cube cube = new Cube(10, 20, 30);

    System.out.printf("\nNew cube: %s%n", cube.print_cube());

    System.out.printf("Width of the cube is : %d%n", cube.width);
    System.out.printf("Height of the cube is : %d%n", cube.height);
    System.out.printf("Depth of the cube is : %d%n", cube.depth);
    System.out.printf("Color of the cube is : %s%n", cube.color);
    System.out.printf("Area of the cube is : %d%n", cube.get_area());
    System.out.printf("Perimeter of the cube is : %d%n", cube.get_perimeter());
    System.out.printf("Volume of the cube is : %d%n", cube.get_volume());
  }
}
